stacks-signer/src/chainstate.rs

```diff
@@ -0,0 +1,375 @@
+    ) -> Result<bool, ClientError> {
+        // if the parent tenure is the last sortition, it is a valid choice.
+        // if the parent tenure is a reorg, then all of the reorged sortitions
+        //  must either have produced zero blocks _or_ produced their first block
+        //  very close to the burn block transition.
+        if sortition_state.prior_sortition == sortition_state.parent_tenure_id {
+            return Ok(true);
+        }
+        info!(
+            "Most recent miner's tenure does not build off the prior sortition, checking if this is valid behavior";
+            "proposed_block_consensus_hash" => %block.header.consensus_hash,
+            "proposed_block_signer_sighash" => %block.header.signer_signature_hash(),
+        );
+
+        let tenures_reorged = client.get_tenure_forking_info(
+            &sortition_state.parent_tenure_id,
+            &sortition_state.prior_sortition,
+        )?;
+        if tenures_reorged.len() == 0 {
+            warn!("Miner is not building off of most recent tenure, but stacks node was unable to return information about the relevant sortitions. Marking miner invalid.";
+                    "proposed_block_consensus_hash" => %block.header.consensus_hash,
+                    "proposed_block_signer_sighash" => %block.header.signer_signature_hash(),
+            );
+            return Ok(false);
+        }
+        for tenure in tenures_reorged.iter() {
+            if tenure.first_block_mined.is_some() {
+                // TODO: must check if the first block was poorly timed.
+                warn!(
+                    "Miner is not building off of most recent tenure, but a tenure they attempted to reorg has already mined blocks.";
+                    "proposed_block_consensus_hash" => %block.header.consensus_hash,
+                    "proposed_block_signer_sighash" => %block.header.signer_signature_hash(),
+                    "parent_tenure" => %sortition_state.parent_tenure_id,
+                    "last_sortition" => %sortition_state.prior_sortition,
+                    "violating_tenure_id" => %tenure.consensus_hash,
+                    "violating_tenure_first_block_id" => ?tenure.first_block_mined,
+                );
+                return Ok(false);
+            }
+        }
+
+        return Ok(true);
+    }
+
+    fn check_tenure_change_block_confirmation(
+        tenure_change: &TenureChangePayload,
+        block: &NakamotoBlock,
+        signer_db: &SignerDb,
+    ) -> Result<bool, ClientError> {
+        // in tenure changes, we need to check:
+        // (1) if the tenure change confirms the expected parent block (i.e.,
+        //     the last block we signed in the parent tenure)
+        // (2) if the parent tenure was a valid choice
+        Self::confirms_known_blocks_in(block, &tenure_change.prev_tenure_consensus_hash, signer_db)
+    }
+
+    fn confirms_known_blocks_in(
+        block: &NakamotoBlock,
+        tenure: &ConsensusHash,
```

Comment:
Is `tenure` necessary?  Is it ever different from `block.header.consensus_hash`?

Reply:
Yes -- for the first block in a new tenure, this method is invoked with the `tenure = &tenure_change.prev_tenure_consensus_hash`